source/material_model/viscoelastic.cc

```diff
@@ -0,0 +1,507 @@
+                           ?
+                           this->get_timestep()
+                           :
+                           fixed_elastic_time_step * year_in_seconds );
+
+      for (unsigned int i=0; i < in.temperature.size(); ++i)
+        {
+          const double temperature = in.temperature[i];
+          const std::vector<double> composition = in.composition[i];
+          const std::vector<double> volume_fractions = compute_volume_fractions(composition);
+
+          out.specific_heat[i] = average_value ( volume_fractions, specific_heats, arithmetic);
+
+
+          // Arithmetic averaging of thermal conductivities
+          // This may not be strictly the most reasonable thing, but for most Earth materials we hope
+          // that they do not vary so much that it is a big problem.
+          out.thermal_conductivities[i] = average_value ( volume_fractions, thermal_conductivities, arithmetic);
+
+          double density = 0.0;
+          for (unsigned int j=0; j < volume_fractions.size(); ++j)
+            {
+              // not strictly correct if thermal expansivities are different, since we are interpreting
+              // these compositions as volume fractions, but the error introduced should not be too bad.
+              const double temperature_factor= (1.0 - thermal_expansivities[j] * (temperature - reference_T));
+              density += volume_fractions[j] * densities[j] * temperature_factor;
+            }
+          out.densities[i] = density;
+
+          out.thermal_expansion_coefficients[i] = average_value ( volume_fractions, thermal_expansivities, arithmetic);
+
+          // Compressibility at the given positions.
+          // The compressibility is given as
+          // $\frac 1\rho \frac{\partial\rho}{\partial p}$.
+          // (here we use an incompressible medium)
+          out.compressibilities[i] = 0.0;
+          // Pressure derivative of entropy at the given positions.
+          out.entropy_derivative_pressure[i] = 0.0;
+          // Temperature derivative of entropy at the given positions.
+          out.entropy_derivative_temperature[i] = 0.0;
+          // Change in composition due to chemical reactions at the
+          // given positions. The term reaction_terms[i][c] is the
+          // change in compositional field c at point i.
+          for (unsigned int c=0; c<in.composition[i].size(); ++c)
+            out.reaction_terms[i][c] = 0.0;
+
+          // Compute average values of viscosity and elastic shear modulus
+          const double vis_avg = average_value ( volume_fractions, viscosities, viscosity_averaging);
+          const double esm_avg = average_value ( volume_fractions, elastic_shear_moduli, viscosity_averaging);
+
+          //Compute viscoelastic (e.g., effective) viscosity (equation 28 in Moresi et al., 2003, J. Comp. Phys.)
+          const double vev_avg = ( vis_avg * dte ) / ( dte + ( vis_avg / esm_avg ) );
```

Comment:
Hmm, I honestly had not considered that someone might use this material model with elasticity. I think your suggestion for returning the average of the regular viscosities makes sense if elasticity is not enabled.

I also agree that it would be best not calculate the effective viscosities more than once. This calculation will need to go in a function regardless, as the particle routines will need to access the effective viscosity value in order to update the stresses.